Telemetry - first pull

asmialek opened the pull request:
I have finished the first version of `telemetry` module, along with `downlink` and `downlink_queue` submodules. There are still some stuff to do, mainly the things mentioned in **`TODO`**s at the top of the files. But I think it's a high time to have first version of it in the `master` branch, for you to see how it's working and complain about every mistake that I did. Don't hesitate to ask questions and have me explain stuff if I wrote spaghetti code or didn't comment it.

asmialek committed:
Initial commit for `telemetry` module

asmialek committed:
Init commit for `downlik_queue` submodule

asmialek committed:
Finish `downlink_queue', implement test in `telemetry.c`

asmialek committed:
Finish `downlink_queue`, telemetry test

asmialek committed:
Finish core of `telemetry` module

asmialek committed:
Restore Makefile

asmialek commented:
If you have any comments or questions that target specific piece of code directly, please go to `Files changed` on this page and add comment with the `+` button next to a line - it helps a lot.

HaraldMagnusson commented on src/init/init.c:
```diff
@@ -29,6 +29,9 @@
 #include "watchdog.h"
 
+//todo: delet this!
+#include "downlink.h"
```
Was this supposed to be included in the pull?

HaraldMagnusson commented on src/telemetry/downlink/downlink.c:
```diff
@@ -7,8 +7,90 @@
+ * I have decided not to put this in `init_downlink()` for the function name to stay
+ * as much related to wwhat happens inside as possible and to make it more abstract.
+ * Just a quick digression and a heads up for further "standards", for anyone who
+ * reads these comments.
```
imo, this is unecessary as the downlink is the udp socket.

HaraldMagnusson commented on src/telemetry/downlink/downlink.c:
```diff
@@ -1,14 +1,85 @@
+static struct sockaddr_in UDP_SERVER;
+
+/* prototypes declaration */
+int init_upd_socket(void);
```
typo: upd -> udp (for all appearances of this function)

HaraldMagnusson commented on src/telemetry/telemetry.c:
```diff
@@ -1,13 +1,22 @@
-/* -----------------------------------------------------------------------------
+/* ------------------------------------------------------------------------------
```
```suggestion
/* -----------------------------------------------------------------------------
```

HaraldMagnusson commented on src/telemetry/downlink_queue/downlink_queue.c:
```diff
@@ -1,14 +1,172 @@
-int init_downlink_queue( void ){
+#include "downlink_queue.h"
+
+pthread_mutex_t downlink_mutex = PTHREAD_MUTEX_INITIALIZER;
```
Though i dont think this is a problem in this case, it is recommended to have all initialization in the init function.

asmialek replied:
I don't see what I could move here, as I don't want to pass around mutexes and conditions as function params. Maybe I'm missing something here

HaraldMagnusson replied:
For every `PTHREAD_HAX_INITIALIZER` there is a `pthread_hax_init()` function where giving `NULL` as the attr argument will have the same effect as the initializer. You can see `watchdog.c` for similar usage for the thread attributes. Here is the man page for the function: http://man7.org/linux/man-pages/man3/pthread_mutex_destroy.3p.html
Hope that helps

asmialek replied:
Ok, thanks! I'll do that

HaraldMagnusson commented on src/telemetry/downlink_queue/downlink_queue.c:
```diff
@@ -1,14 +1,172 @@
+        // Special Case: The head of list has lesser
+        // priority than new node. So insert new
+        // node before head node and change head node.
+        if ((*head)->priority > p) {
```
Should it not be like this?
```suggestion
        if ((*head)->priority < p) {
```

asmialek replied:
The lower the `p`, the higher the priority. So `0` is the highest one. `head` is the existing node, so if it has lower priority (so higher number)  than `p` of the data being added, the new node is added as the first. Hence `if head.priority > p`.

HaraldMagnusson commented on src/telemetry/telemetry.c:
```diff
@@ -16,19 +25,59 @@
 
 #define MODULE_COUNT 2
 
+static downlink_node *downlink_queue = NULL;
```
I think it would be better for this to be local in downlink_queue, and have the `read_downlink_queue` and `send_telemetry_local` not take the queue as an argument.

asmialek replied:
It's the remnant of a previous version. Coming in the next commit.

HaraldMagnusson commented on src/telemetry/telemetry.h:
```diff
@@ -9,7 +9,4 @@
-int init_telemetry( void );
-
-/* queue up a message to be sent to ground */
-void send_telemetry( void );
```
The design idea here is that only the `telemetry` module is called from outside and not its children. `telemetry` would therefore need to delegate the "work" down to its children through functions like this. This is to avoid unnecessary conflicts between local functions such as `read_queue` in `downlink_queue` and in `img_processing/data_queue` since these would only be included in their respective parents and therefore never together.

asmialek replied:
Ok, now I see the reason for the design. I agree to all that you wrote, I'll include it in the next commit.

asmialek replied:
I didn't like the idea of having something like:
```c
int send_telemetry_local(int d, int p) {
    push(&downlink_queue, d, p);
    return SUCCESS;
}
```
in `downlink_queue.c` and then:
```c
int send_telemetry( d, p ) {
    return send_telemetry_local(d, p);
}
```
in `telemetry.c`, as it seems redundant, but I can agree that it helps with avoiding naming and import conflicts.

HaraldMagnusson replied:
Perhaps you can either rename `push` to `send_telemetry_local` or replace the latter with the first to avoid one step. Either way, the function in `downlink_queue.c` won't be accessible from outside of `telemetry` and its children.

HaraldMagnusson commented on Makefile:
```diff
@@ -9,4 +9,4 @@ clean:
 
 degen:
-	rm build/* -rf
+	rm build/* -rf
```
git gets a bit mad if you remove newlines :c

asmialek replied:
I'm adding a blank line at the bottom, if that's what you mean.

HaraldMagnusson commented on src/telemetry/telemetry.c:
```diff
@@ -16,19 +25,59 @@
+ * @TODO Move it to `init_telemetry`?
+ * @TODO Redo it all!!!
+ */
+void *telemetry_sender() {
```
Designwise, it was intented that what this thread does would be the responsibility of the `downlink` module.

asmialek replied:
As you wrote, I'll move contents of `downlink` into `e-link` and I can move this into `downlink` then. I'll keep this convo open, to remember about it.

HaraldMagnusson commented on src/telemetry/downlink/downlink.c:
```diff
@@ -1,14 +1,85 @@
+ *
+ * @return
+ */
+int init_upd_socket(void) {
```
It was intended that the `e_link` module should handle the ip communications. Note that the declarations in `e_link.h` are preliminary and can be changed. Currently, the contents of this file do not match the purpose of the module.

HaraldMagnusson commented on src/telemetry/telemetry.c:
```diff
@@ -16,19 +25,59 @@
 
-int init_telemetry( void ){
+// TODO: For test purposes only!
+void *test_telemetry_generator() {
```
I think we need to design a system for code testing. Perhaps for now this should be removed to not clog up the production code with testing?

asmialek replied:
Yeah, I'll keep the code itself for now, but I'll comment it out, so it's not clogging the workflow of the daemon.

HaraldMagnusson reviewed (changes requested):
Please see the comments in the code.

HaraldMagnusson committed:
Update src/telemetry/telemetry.c

Co-Authored-By: asmialek <smialekadam@gmail.com>

asmialek committed:
WIP: Refactor telemetry module
I was doing something here, but I want to work on another branch, so
just in case I'll commit. I hope I won't brake anything, fingers crossed

WEriksson committed:
Telemetry WIP, merging master for logging

WEriksson committed:
Merge branch 'master' into telemetry. Merged master for logging

WEriksson committed:
WIP, TM/TC working, still need to implement some of the commands
